docs/amdgpu_kernel_optimization_guide.md

```diff
@@ -0,0 +1,282 @@
+| N/A | (Thread) Block Cluster | N/A | Shader Engine |
+| (Work)group counts | Grid | Grid | GPU |
+| Workgroup Memory | Shared Memory | Local Data Store | LDS & Crossbar (Compute Unit) |
+
+The rest of the document uses the Vulkan and AMDGPU naming for consistency.
+
+## GFX9 Architecture Overview
+
+The GFX9 line of hardware covers both consumer Radeon GPUs (GCN) and newer
+datacenter MI accelerators like MI300 (CDNA3).
+
+### MI300X Compute Topology
+
+The MI300 GPU uses a chiplet design. For example, the MI300X variant consists
+of 8 chiplets called XCDs, sat on 4 pairs of AIDs/IODs. One XCD contains 4
+Shader Engines with the total of 38 Compute Units (CUs). One CU contains 4
+16-lane-wide SIMDs. In total, MI300X has 304 CUs.
+
+> [!TIP]
+> The number of workgroups launched should be a multiple of the number of CUs to
+> continuously utilize the whole GPU.
+
+A kernel is dispatched to one or more CUs.
+
+![MI300 Topology](./assets/mi300_topology.png)
+
+![MI300 IOD](./assets/mi300_iod.png)
+
+> Source: [The CDNA3
+> Whitepaper](https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/white-papers/amd-cdna-3-white-paper.pdf)
+
+Each XCD has its own L2 cache and is connected to LLC cache (a.k.a. MALL:
+memory-attached last-level, a.k.a 'Infinity Cache') via its IOD.
+
+### MI300X Cache Hierarchy
+
+There are 3 levels of cache:
+
+| Name | Size | Cache Line Size | Associativity | Execution Unit | Comments |
+| --- | --- | --- | --- | --- | --- |
+| L1D | 32 kB | 128 B | 64-way, 4 sets | Compute Unit | Write-through |
+| L1I | 64 kB | 128 B | 8-way set-associative | Compute Unit | Instruction cache |
+| L2 | 4 MB (16 channels * 256 kB) | 128 B | 16-way set-associative | XCD | Writeback / Write-allocate, Coherent within XCD |
+| LLC | 32 MB (16 channels * 2 MB), 256 MB total | 64 B | 16-way set-associative | IOD | Coherent, MALL |
+
+L2 cache is flushed between kernel launches. Memory accesses that miss L2 are
+coalesced and go to the data fabric.
+
+> [!TIP]
+> Due to power consumption, we want to minimize the number of data fabric
+> transactions.
+
+### Execution Model
+
+When a kernel is launched, its workgroups get distributed across the GPU.
+A workgroup executes on a single CU and never gets migrated to another CU.
+
+Each subgroup / wave within the workgroup gets assigned to a single SIMD unit.
+One SIMD has 10 waveslots used to 'context-switch' between the assigned subgroups
+(up to 10). However, only up to 16 subgroups are allowed within a single
+workgroup.
+
+On GFX9, the subgroup size is 64, which, for most instructions, necessitates
+multiple clock cycles on 16-lane SIMD. For example, something like an `add`
+would execute in 4 cycles, for each set of 16 threads within the subgroup.
+
+> [!TIP]
+> To fully utilize all 4 SIMDs within the CU, use a workgroup size of 256 (or a
+> multiple of 256). To conserve power, you can use workgroup size 128 so that 2
```

Comment:
Writing "use a workgroup size of 128" might be more consistent with the previous sentence.